Cruise Control deployment (#2773)
strimzi/strimzi-kafka-operator: documentation/modules/cruise-control/ref-cruise-control-configuration.adoc

// Module included in the following assemblies:
//
// assembly-cruise-control-concepts.adoc

[id='ref-cruise-control-configuration-{context}']
= Cruise Control configuration

The `config` property in `Kafka.spec.cruiseControl` contains configuration options as keys with values as one of the following JSON types:

* String
* Number

> * Boolean

Reviewer: This applies more generally to the rest of the docs too, but here would be a good place for a `NOTE` to say that if your String happens to look like YAML or JSON it will need to be explicitly quoted.

* `bootstrap.servers`
* `zookeeper.`
* `ssl.`
* `security.`
* `failed.brokers.zk.path`
* `webserver.http.port`
* `webserver.http.address`
* `webserver.api.urlprefix`
* `metric.reporter.sampler.bootstrap.servers`
* `metric.reporter.topic`
* `metric.reporter.topic.pattern`
* `partition.metric.sample.store.topic`
* `broker.metric.sample.store.topic`
* `capacity.config.file`
* `skip.sample.store.topic.rack.awareness.check`
* `cruise.control.metrics.topic`
* `sasl.`

If restricted options are specified, they are ignored and a warning message is printed to the Cluster Operator log file.
All supported options are passed to Cruise Control.

.An example Cruise Control configuration
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: Kafka
metadata:
  name: my-cluster
spec:
  # ...
  cruiseControl:
    # ...
    config:
      default.goals: >
         com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,
         com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal
      cpu.balance.threshold: 1.1
      metadata.max.age.ms: 300000
      send.buffer.bytes: 131072
    # ...
----
== Capacity configuration

> Cruise Control uses capacity limits to determine if goals related to broker resources are being violated.

Reviewer: Which begs the question: What happens if the limits are violated?

Author: Added some text:

> A violation of a goal can cause a optimization to fail, preventing that optimization from being used to generate a partition rebalance proposal.
For example, an optimization that would cause a broker to exceed its CPU capacity would not be used for when the CpuCapacityGoal is set as a hard goal

Capacity limits for broker resources can be specified in the `capacity` property in `Kafka.spec.cruiseControl`.
Capacity limits can be set for the following broker resources in the described units:

* `disk`  - Disk storage in MB
* `cpu`   - CPU utilization as a percent (0-100)
* `networkIn`  - Network inbound throughput in KB/s
* `networkOut` - Newtork outbound throughput in KB/s

> Since {ProductName} brokers are homogenous, Cruise Control will apply the same capacity limits to every broker it is monitoring.

Reviewer: ```suggestion
Since {ProductName} brokers are homogeneous, Cruise Control will apply the same capacity limits to every broker it is monitoring.
```

// Module included in the following assemblies:
//
// assembly-cruise-control-concepts.adoc

[id='ref-cruise-control-configuration-{context}']
= Cruise Control configuration

The `config` property in `Kafka.spec.cruiseControl` contains configuration options as keys with values as one of the following JSON types:

* String
* Number
* Boolean

NOTE: Strings that look like JSON or YAML will need to be explicitly quoted.

You can specify and configure all the options listed in the "Configurations" section of the {CruiseControlConfigDocs}, apart from those managed directly by {ProductName}.
Specifically, you *cannot* modify configuration options with keys equal to or starting with one of the following strings:

* `bootstrap.servers`
* `zookeeper.`
* `ssl.`
* `security.`
* `failed.brokers.zk.path`
* `webserver.http.port`
* `webserver.http.address`
* `webserver.api.urlprefix`
* `metric.reporter.sampler.bootstrap.servers`
* `metric.reporter.topic`
* `metric.reporter.topic.pattern`
* `partition.metric.sample.store.topic`
* `broker.metric.sample.store.topic`
* `capacity.config.file`
* `skip.sample.store.topic.rack.awareness.check`
* `cruise.control.metrics.topic`
* `sasl.`

> 

Reviewer: We'll need to update this for the ssl ones too.

Author: I am going to enable ssl configurations until I secure the communication between Cruise Control and Kafka brokers with SSL

Reviewer: OK

If restricted options are specified, they are ignored and a warning message is printed to the Cluster Operator log file.
All supported options are passed to Cruise Control.

.An example Cruise Control configuration
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: Kafka
metadata:
  name: my-cluster
spec:
  # ...
  cruiseControl:
    # ...
    config:
      default.goals: >
         com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,
         com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal
      cpu.balance.threshold: 1.1
      metadata.max.age.ms: 300000
      send.buffer.bytes: 131072
    # ...
----
== Capacity configuration

> Cruise Control uses capacity limits to determine if goals related to broker resources are being violated.

Reviewer 2: ```suggestion
Cruise Control uses _capacity limits_ to determine if optimization goals for broker resources are being broken.
```

> A violation of a goal can cause a optimization to fail, preventing that optimization from being used to generate a partition rebalance proposal.

Reviewer 2: ```suggestion
An optimization will fail if a goal is broken, preventing the optimization from being used to generate an optimization proposal for balanced partition assignments.
```

> For example, an optimization that would cause a broker to exceed its CPU capacity would not be used for when the CpuCapacityGoal is set as a hard goal.

Reviewer 2: ```suggestion
For example, an optimization that would cause a broker to exceed its CPU capacity would not be used if the `CpuCapacityGoal` is set as a hard goal.
```

> Capacity limits for broker resources can be specified in the `brokerCapacity` property in `Kafka.spec.cruiseControl`.

Reviewer 2: ```suggestion
You specify capacity limits for broker resources in the `brokerCapacity` property in `Kafka.spec.cruiseControl` .
```

Capacity limits can be set for the following broker resources in the described units:

* `diskMiB`  - Disk storage in MiB
* `cpuUtilization`   - CPU utilization as a percent (0-100)
* `inboundNetworkKiBPerSecond`  - Network inbound throughput in KiB/s

> * `outboundNetworkKiBPerSecond` - Newtork outbound throughput in KiB/s

Reviewer 2: ```suggestion
* `outboundNetworkKiBPerSecond` - Network outbound throughput in KiB/s
```

> Since {ProductName} brokers are homogeneous, Cruise Control will apply the same capacity limits to every broker it is monitoring.

Reviewer 2: ```suggestion
Because Kafka brokers are homogeneous, Cruise Control applies the same capacity limits to every broker it is monitoring.
```

Author: Updated to {ProductName} Kafka brokers since Kafka brokers are not necessarily homogeneous but Kafka brokers managed by Strimzi are homogeneous

.An example Cruise Control brokerCapacity configuration
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: Kafka
metadata:
  name: my-cluster
spec:
  # ...
  cruiseControl:
    # ...
    brokerCapacity:
      diskMiB: 100000
      cpuUtilization: 100
      inboundNetworkKiBPerSecond: 10000
      outboundNetworkKiBPerSecond: 10000
    # ...
----

.Additional resources

> For information, refer to the xref:type-CruiseControlBrokerCapacity-reference[Capacity schema reference].

Reviewer 2: ```suggestion
For more information, refer to the xref:type-CruiseControlBrokerCapacity-reference[].
```
